### DISCUSSION TITLE
Non-Homogeneous Reaction-Diffusion Equation with Robin BC

### DISCUSSION THREAD

**Discussion by charlotte-b8086** [idaholab/moose#27389] [Q&A General]
I've been trying to understand how Neumann and Robin Boundary Conditions are implemented in MOOSE.

For one example, I decided to code up a very simple Reaction-Diffusion Equation with Robin BC whose solution I know to verify that MOOSE could converge to the correct solution. MOOSE converged quickly, and while the solution appears to have the correct shape, the values are incorrect. 

The PDE I was considering was 
```tex
\begin{equation*}
-\Delta u + u = (\pi^2 + 1)\sin(\pi x) \textrm{ on } (0, 1) \textrm{ subject to } \frac{\partial u}{\partial n} + u = 
\begin{cases}
-\pi \cos(\pi x) + \sin(\pi x) & x = 0 \\
\pi \cos(\pi x) + \sin(\pi x) & x = 1
\end{cases}.\end{equation*}
``` 

Solving this PDE using the Method of Undetermined Coefficients, I found that its solution is $u(x) = \sin(\pi x)$. 

I implemented the Kernels in MOOSE as follows:
```c++
[Kernels]
  #R_u: (nabla u, nabla test) + (beta u, test) + (-f, test) = 0
  #f = (pi^2 + beta)sin(pix)
  [./u_diffusion]
    variable = u
    type = ADDiffusion
  [../]
  [./u_reaction]
    variable = u
    type = ADCoefReaction
  [../]
  [./force]
    variable = u
    type = ADBodyForce
    function = '(pi*pi+1.0)*sin(pi*x)'#force_func
  [../]
[]
``` 
For the Boundary Conditions I made a Custom Moose Object:
```c++
ADNonHomogeneousRobinBC::ADNonHomogeneousRobinBC(const InputParameters & parameters)
  : ADIntegratedBC(parameters),
  _coef(getParam<Real>("gamma")),
  _n(getParam<Real>("normal_dir"))
{
}

ADReal
ADNonHomogeneousRobinBC::computeQpResidual()
{
  return -_coef * _test[_i][_qp]*_u[_qp] +
      _test[_i][_qp]*_n* (libMesh::pi)*std::cos(libMesh::pi * _q_point[_qp](0))+ _coef *_test[_i][_qp]*std::sin(libMesh::pi * _q_point[_qp](0));
}
```
Please note that I have as input in my .i file normal_dir is -1 for the left boundary (x = 0) and normal_dir is 1 for the right boundary (x = 1).

However, Paraview showed my solutions as: 
<img width="519" alt="Screenshot 2024-04-16 at 9 38 10 PM" src="https://github.com/idaholab/moose/assets/112354003/af49749c-9498-405f-b9ad-19f8ae802d17">
Initially I thought that this was precisely the correct solution $\sin(\pi x)$ with a strange vertical translation, but then I realized that the amplitude is larger than should be. While it still has the correct 'shape' of the solution curve, I don't believe that this is a correct solution for the PDE. I made the nonlinear and linear absolute errors incredibly small and there was no change. 
Is there something wrong with my Custom BC object? I am not sure why this is happening.

I wanted to try to code this BC with ADMatNeumannBC as well to see if I could get the expected result that way (this was a recommendation in discussion post #24108 for how to code Robin BC), but I am unsure how to create a space-dependent material object (and vice verse if I were to attempt FunctionNeumannBC a variable-dependent function). 

Thank you for any assistance you can offer.

**Comment by GiudGiud** [comment]
Hello

I cant find a problem except maybe _coeff showing up in the sin(pi x) term in the boundary condition which is not in the equation
but isnt the sin(pi x) 0 on both boundaries?

I would check the implementation of the kernels with Dirichlet boundary conditions and an analytic solution (or with our MMS tools).

Guillaume

**Reply by charlotte-b8086** [reply]
Hello,

Thank you for your response.

There was a theoretical reason why I didn't have the coefficient multiplier on the last term in the computeQpResidual and two practical reasons. For the first practical reason; as you pointed out, $\sin(\pi x)$ is 0 at $x = 0$ and $x = 1$. 
For the theoretical reason, when formulating the problem I began by working with the ODE $-\Delta u + \beta u = f$ on $\Omega$ subject to the boundary condition $\frac{1}{\gamma}\frac{\partial u}{\partial n} + u = g$ on $\partial \Omega$. In implementation, I set $\beta = \gamma = 1$ (the second practical reason), but even without that being 1, rewriting this boundary condition, one has $\frac{\partial u}{\partial n} = \gamma \left( g - u\right)$, and the function $g$ itself, when actually picking it, is defined as $\frac{1}{\gamma} \frac{\partial u}{\partial n} + u$. When multiplying this by $\gamma$, the coefficient on the first term drops out.

When coding this PDE with the Dirichlet BCs $u(0) = 0 = u(1)$ the solution Paraview provides is correct. The convergence also is as expected. 
![mms_spatial](https://github.com/idaholab/moose/assets/112354003/a06c335c-7379-4558-a206-0d94ca1cf349)

The convergence with the BC I developed, ADNonHomogeneousRobinBC, is terrible, which makes sense since, as confirmed by the MMS tool, the solution MOOSE provided is off from the correct solution $\sin(\pi x)$.
![mms_spatial_ADNonHomogeneousRobinBC](https://github.com/idaholab/moose/assets/112354003/b3d5807b-3166-436c-9b40-4e67d8dd4abd)

If you meant I should try to rewrite the Robin boundary condition in such a way as to implement it as a Dirichlet BC, that is an approach I haven't attempted yet. I believe that I would need to include the ADDirichletBCTempl.h file in my header file in order to override computeQpValue(). However, this header file doesn't know the _u_grad parameter. I know that ADIntegratedBC.h does have that as a default parameter. Would there be a way to inherit from both templates? 

Thank you for the suggestions and assistance.

**Reply by GiudGiud** [reply]
> If you meant I should try to rewrite the Robin boundary condition in such a way as to implement it as a Dirichlet BC, that is an approach I haven't attempted yet. I believe that I would need to include the ADDirichletBCTempl.h file in my header file in order to override computeQpValue(). However, this header file doesn't know the _u_grad parameter. I know that ADIntegratedBC.h does have that as a default parameter. Would there be a way to inherit from both templates?

I didnt. you wont be able to inherit both classes, you would get diamond inheritance issues. 

If you did want to do that, you will want to try using the Functor interfaces on variables.
Create a new class, and retrieve a pointer to the actual variable (just _var maybe?), then you can do _var->grad(...)
see the doxygen for variables. 

What reference are you using for the MMS study? The analytical solution or the most converged case?

**Reply by charlotte-b8086** [reply]
I was trying to follow the [MMS documentation](https://mooseframework.inl.gov/python/mms.html). If I understand it correctly, it was using the analytical solution as reference, since this is what is spit out by 
```c++
fs,ss = mms.evaluate('-div(grad(u))', 'sin(2*pi*x)*sin(2*pi*y)')
mms.print_fparser(fs)
```
